symFileManager.py

```diff
@@ -1,232 +1,69 @@
-import time
-import urllib2
-import urlparse
-import gzip
-from bisect import bisect
-from StringIO import StringIO
-from tempfile import NamedTemporaryFile
-
-class SymbolInfo:
-  def __init__(self, addressMap):
-    self.sortedAddresses = sorted(addressMap.keys())
-    self.sortedSymbols = [addressMap[address] for address in self.sortedAddresses]
-    self.entryCount = len(self.sortedAddresses)
-
-  # TODO: Add checks for address < funcEnd ?
-  def Lookup(self, address):
-    nearest = bisect(self.sortedAddresses, address) - 1
-    if nearest < 0:
-      return None
-    return self.sortedSymbols[nearest]
-
-  def GetEntryCount(self):
-    return self.entryCount
+from symLogging import LogDebug, LogMessage
+from symFetcher import PathFetcher, URLFetcher
+from symCache import MemoryCache, DiskCache
 
 # Singleton for .SYM file cache management
 class SymFileManager:
-  # Symbol cache data structures
-  sCache = {}
-  sCacheCount = 0
-  sMruSymbols = []
-  sUpdateMRUFile = True
-
-  sOptions = {}
-
   def __init__(self, options):
     self.sOptions = options
 
-  def GetLibSymbolMap(self, libName, breakpadId):
-    CheckDebug()
+    self.fetchPipeline = (PathFetcher(options), URLFetcher(options))
+    self.memoryCache = MemoryCache(options)
+    self.diskCache = DiskCache(options)
+    assert self.memoryCache.MAX_SIZE <= self.diskCache.MAX_SIZE
 
-    # Empty lib name means client couldn't associate frame with any lib
-    if libName == "":
-      return None
+    self.MRU = self.diskCache.GetCacheEntries()
+    self.memoryCache.LoadCacheEntries(self.MRU, self.diskCache)
 
-    # Check cache first
-    libSymbolMap = None
-    if libName in self.sCache and breakpadId in self.sCache[libName]:
-      libSymbolMap = self.sCache[libName][breakpadId]
-      self.UpdateMruList(libName, breakpadId)
+    LogMessage("MRU loaded with {} entries".format(len(self.MRU)))
 
-    if libSymbolMap is None:
-      LogDebug("Need to fetch PDB file for " + libName + " " + breakpadId)
+  def GetLibSymbolMap(self, lib):
+    # Empty lib name means client couldn't associate frame with any lib
+    if lib[0] == "":
+      return None
 
-      # Guess the name of the .sym file on disk
-      if libName[-4:] == ".pdb":
-        symFileName = re.sub(r"\.[^\.]+$", ".sym", libName)
+    try:
+      index = self.MRU.index(lib)
+      if index < self.memoryCache.MAX_SIZE:
+        cache = self.memoryCache
       else:
-        symFileName = libName + ".sym"
-
-      pathSuffix = os.path.join(libName, breakpadId, symFileName)
-      urlSuffix = "/".join([libName, breakpadId, symFileName])
-
-      # Look in the symbol dirs for this .sym file
-      for symbolPath in self.sOptions["symbolPaths"]:
-        path = os.path.join(symbolPath, pathSuffix)
-        libSymbolMap = self.FetchSymbolsFromFile(path)
+        cache = self.diskCache
+      LogDebug("Loading [{}] [{}] from {}".format(lib[0], lib[1], cache.__class__))
+      libSymbolMap = cache.Get(lib)
+    except ValueError:
```

Comment:
doesn't DiskCache.Get() return None if it can't find the file on disk? that means ValueError wouldn't get thrown, right?